sphinx/ext/collapse.py

```diff
@@ -0,0 +1,206 @@
+    else:
+        translator.body.append(translator.starttag(node, 'details'))
+
+
+def depart_collapsible(translator: HTML5Translator, node: nodes.Element) -> None:
+    translator.body.append('</details>\n')
+
+
+def visit_summary(translator: HTML5Translator, node: nodes.Element) -> None:
+    translator.body.append(translator.starttag(node, 'summary'))
+
+
+def depart_summary(translator: HTML5Translator, node: nodes.Element) -> None:
+    translator.body.append('</summary>\n')
+
+
+class Collapsible(SphinxDirective):
+    """
+    Directive to mark collapsible content, with an optional summary line.
+    """
+
+    has_content = True
+    optional_arguments = 1
+    final_argument_whitespace = True
+    option_spec: ClassVar[OptionSpec] = {
+        'class': directives.class_option,
+        'name': directives.unchanged,
+        'open': directives.flag,
+    }
+
+    def run(self) -> list[nodes.Node]:
+        node = collapsible(
+            classes=self.options.get('classes', []),
+            collapsible_open='open' in self.options,
+        )
+        self.add_name(node)
+        node.document = self.state.document
+        self.set_source_info(node)
+
+        if len(self.arguments) > 0:
+            # parse the argument as reST
+            trimmed_summary = self._prepare_argument_string(self.arguments[0].strip())
+            textnodes, messages = self.state.inline_text(trimmed_summary, self.lineno)
+            node.append(summary(trimmed_summary, '', *textnodes))
+            node += messages
+        else:
+            label = 'Collapsed Content:'
+            node.append(summary(label, label))
+
+        return self._parse_content(self.state, node, self.content, self.content_offset)
+
+    @staticmethod
+    def _prepare_argument_string(s: str) -> str:
+        """Prepare a directive argument string.
+
+        Remove common leading indentation, where the indentation of the first
+        line is ignored.
+
+        Return a list of lines usable for inserting into a docutils StringList.
+        """
+        lines = s.expandtabs().splitlines()
+
+        # Find minimum indentation of any non-blank lines after the first.
+        # If your indent is larger than a million spaces, there's a problem…
+        margin = 10**6
+        for line in lines[1:]:
+            content = len(line.lstrip())
+            if content:
+                indent = len(line) - content
+                margin = min(margin, indent)
+
+        if margin == 10**6:
+            return s
+
+        return '\n'.join(lines[:1] + [line[margin:] for line in lines[1:]])
+
+    @staticmethod
+    def _parse_content(
+        state: RSTState,
+        node: nodes.Element,
+        content: StringList,
+        offset: int,
+    ) -> list[nodes.Node]:
+        # Same as util.nested_parse_with_titles but try to handle nested
+        # sections which should be raised higher up the doctree.
```

Comment:
I have to say I'm completely against this kind of "hacking" into docutils.
If you want to do something like this, then either it needs to be upstreamed to docutils, or it needs much, much more testing than is in this PR.
simple example: what happens if you put your `collapse` directive inside a `note` directive?

Not only is it already brittle in rst, but also its definitely not supported by myst 😅

Reply:
I certainly get the desire to have nested parsing with sections (in fact I am trying to achieve it in https://sphinx-rust.readthedocs.io)

but this should be a "core" function within sphinx/docutils, with good testing, rather than having multiple implementations scattered across the code base